### PR TITLE
Assisted inject for datanode migration context

### PR DESCRIPTION
**Pull request by todvora**
The existing MigrationActionsImpl was a singleton. We were setting the MigrationStateMachineContext to a singleton from different request threads, losing anything that the other thread could have written to the context before. This occured during parallel /state and /trigger request during migrations.

New implementation is using a factory for the MigrationActions, creating new instance for each request, propagating a new instance of MigrationStateMachineContext for each thread. This separates the actions contexts of several threads, never mixing them.

/nocl

## Motivation and Context
Sometimes during the connection check in remote reindex migration, a message "No indexes available" has been displayed. There was no problem with the connection check itself, but due to the threading issue and context overriding, we lost the check info and never returned it. 

## How Has This Been Tested?
Manually + existing set of unit and integration tests. 

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Refactoring (non-breaking change)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.

### CODE REVIEW

File: graylog2-server/src/main/java/org/graylog/plugins/views/storage/migration/state/actions/MigrationActions.java [hunk 1]
```diff
@@ -55,10 +55,6 @@ public interface MigrationActions {
 
     boolean dataNodeStartupFinished();
 
-    void setStateMachineContext(MigrationStateMachineContext context);
```
Comment (moesterheld) [inline graylog2-server/src/main/java/org/graylog/plugins/views/storage/migration/state/actions/MigrationActions.java:58]: unused import of `MigrationStateMachineContext` should be removed

File: graylog2-server/src/main/java/org/graylog/plugins/views/storage/migration/state/machine/MigrationStateMachineProvider.java [hunk 2]
```diff
@@ -21,23 +21,26 @@
...
 @Singleton
 public class MigrationStateMachineProvider implements Provider<MigrationStateMachine> {
 
     private final DatanodeMigrationPersistence persistenceService;
-    private final MigrationActions migrationActions;
+    private final MigrationActionsFactory migrationActionsFactory;
 
     @Inject
-    public MigrationStateMachineProvider(DatanodeMigrationPersistence persistenceService, MigrationActions migrationActions) {
+    public MigrationStateMachineProvider(DatanodeMigrationPersistence persistenceService, MigrationActionsFactory migrationActions) {
```
Comment (moesterheld) [inline graylog2-server/src/main/java/org/graylog/plugins/views/storage/migration/state/machine/MigrationStateMachineProvider.java:34]:
```suggestion
    public MigrationStateMachineProvider(DatanodeMigrationPersistence persistenceService, MigrationActionsFactory migrationActionsFactory) {
```
rename var

File: graylog2-server/src/main/java/org/graylog/plugins/views/storage/migration/state/machine/MigrationStateMachineProvider.java [hunk 3]
```diff
@@ -21,23 +21,26 @@
...
        this.persistenceService = persistenceService;
-       this.migrationActions = migrationActions;
+       this.migrationActionsFactory = migrationActions;
```
Comment (moesterheld) [inline graylog2-server/src/main/java/org/graylog/plugins/views/storage/migration/state/machine/MigrationStateMachineProvider.java:36]:
```suggestion
       this.migrationActionsFactory = migrationActionsFactory;
```
rename var

**Review by moesterheld — APPROVED** [review]
Very nice solution. just some minor cosmetic changes.

File: graylog2-server/src/main/java/org/graylog/plugins/views/storage/migration/state/actions/MigrationActions.java (hunk 1 above)
Comment (todvora) [reply graylog2-server/src/main/java/org/graylog/plugins/views/storage/migration/state/actions/MigrationActions.java:58]: removed, thx!

File: graylog2-server/src/main/java/org/graylog/plugins/views/storage/migration/state/machine/MigrationStateMachineProvider.java (hunk 3 above)
Comment (todvora) [reply graylog2-server/src/main/java/org/graylog/plugins/views/storage/migration/state/machine/MigrationStateMachineProvider.java]: fixed, thx!

**Pull request merged**